### DISCUSSION TITLE
Does a path always exist to the sender of an announce?

### DISCUSSION THREAD

**Discussion by ThomasFreedman** [markqvist/Reticulum#311] [Help & Questions] [answered]
*Mod edit: This is a series of important questions, that include some common misconceptions. The marked answer tries to clear those up. If something is still unclear, please post it, and your questions shall be answered :)*

-----

I'm still a bit confused about paths.

I have successfully setup a link in response to my app's announcement. Lets call the app the "server".  A link is created by a responder to the announcement. In the responder's announce handler RNS.Transport.has_path(destination_hash) always returns True in my testing thus far.

When the responder creates the link it checks if a path exists and sends a request for one if not. A path must exist to the destination to create the link.

I setup a callback _on the server_ which prints an acknowledgement to the link the responsder creates, but I don't see any requests on the server tho the responder is sending them.

Q1: Is the path request code really necessary to setup a link to the source of an announcement, or will a path to the announcer always exist?

Q2: What path value do I provide on the server to create a request callback handler?

Registering a request handler on the server app in the link established callback seems like a good way to do it, but the API call requires a path value and I'm not sure how to get it. 

I have looked at the Link and Destination objects and don't see a "path" property. I used a random string as in the request/response example but that's where my confusion comes in, as a random string doesn't make sense to me in light of the explanation of path being a dot delimited string. It seems that "path" has 2 different contextual meanings.

The only additional aspect filter I'm using is the app_name. An app_name is required to create a Destination, so including any additional aspect values seem redundant, but not sure if it is (I only want to see announcements from instances of my app).

I tried using the app_name as the path and that didn't work either.  

I also noticed the link_closed handler isn't being called on the server, even when I exit the app. I do see it on the responder, but it's not immediate, and is probably due to a timeout occuring.

**Comment by cobraPA** [comment]
Have you seen this document yet: https://reticulum.network/manual/Reticulum%20Manual.pdf ?  I can't answer your question directly, but just ran across it yesterday and it goes in to the network layer in quite a bit of detail.  I am learning about how announces work and had the network layer somewhat confused with LXMF, having looked at Sideband a lot first.  Apologies if you've already seen it.

**Reply by ThomasFreedman** [reply]
No I have not, it's hot off the press! I suspect Mark has been working on that for some time and saw the recent uptick in activity here and published it to save him time to answer questions.

Thank you so much!!!

**Comment by ThomasFreedman** [comment]
After looking it over it appears to be a PDF of what is already available online at: https://markqvist.github.io/Reticulum/manual/reference.html. It is slightly newer though. I need to study it a bit more, but my cursory examination indicates the questions above still remain. On page 31 of the PDF Mark states:

_With this information, any Reticulum node that receives it will be able to reconstruct an outgoing destination to securely communicate with that destination.
...
The use of app names and aspects makes it easy to structure Reticulum programs and makes it possible to filter what information and data your program receives._

I'm only using one aspect element (app_name). His statement leads me to believe a path will always exist to the source of an announce message, but I'm looking for confirmation of that.

Also, I would like to know how to obtain the path (if one exists) from the API, so I can use it to create a register_request_handler. 

It is not clear to me where the path ends and aspects begin. My app name is the only aspect I use, which I believe is why I only see announcements from my app server. Yet, if I only pass my app name as the path parameter of the messages the responder sends, I don't see any messages, despite the server showing the link between it and responder was created.

**Comment by faragher** [comment]
The documentation has some weak spots, and aspects are one of them.

My understanding is it's similar to an FQDN but functionally a filter. Think of it like how a router only sends packets it's interested in, while a switch sends everything. I think. So the app name is required, and one or more aspects are also required. These aspects must match for communication. That has bitten me so many times in learning the system. Also, Path is super overloaded as a term.

Obtaining a path can be done through request_path:

- RNS.Transport.request_path(destination_hash)
- RNS.Transport.request_path(destination_hash, on_interface=interface, tag=tag, recursive=True)

There is slightly more information on page 144 of the manual. 

path_request is also possible, but it's not listed in the API and is basically for internal Transport use.:
- RNS.Transport.path_request(destination_hash, is_from_local_client, attached_interface)
- RNS.Transport.path_request(destination_hash, is_from_local_client, attached_interface, requestor_transport_id=None, tag=None)

It's an actual path, a route through the network to a destination. I don't know the mechanics, but since the network is dynamic, it's actually a request to various (distribution?) nodes to see if anyone knows where to send the packet for the next hop. It's basically querying the network for a cached announce (which isn't technically true, since an announce propagates, generating a path, but still)

I would suggest using the link.py example as a first step:

```
    # Check if we know a path to the destination
    if not RNS.Transport.has_path(destination_hash):
        RNS.log("Destination is not yet known. Requesting path and waiting for announce to arrive...")
        RNS.Transport.request_path(destination_hash)
        while not RNS.Transport.has_path(destination_hash):
            time.sleep(0.1)
```

So a path is the route, and the app_name+aspects is more a naming/filtering system. This is to the best of my knowledge.

**Reply by ThomasFreedman** [reply]
Thank you for your reply.  You can request the transport obj to find a path, or ask it if a path exists, but there's no API to ask what the path is. Even if path_request is an accessible property, I wouldn't want to rely on it, tho it might be useful for debugging.

I'm now realizing when the API has a path argument, it's not the _full path_ (all the hops to get to the destination), but rather it's just some portion of it, which as yo point out, is not made clear in the docs.

I presume app_name, the last required parameter to create a Destinaltion object prior to aspects is inherently part of the path. I also include it as the only aspect, and I only use app_name as the value for "path" when the API requires one. However I still get nothing on the server. There are no errors, nothing in the logs on either server or responder after the link is created on responder and I get the notice of it on the server.

**Reply by faragher** [reply]
Apologies, I don't know the *content* of the path. We're well past my understanding of the network stack. The destination name (such as returned from Destination.Expand_Name) is a period delimited string with app_name.aspect[0].aspect[1].aspect[n] but I haven't ported the path information to C#, and Python's untyped variables is a liability here. 

Transport.path_request_handler has some comments that you could use to reverse engineer the format, but we're probably going to have to wait for Mark.

**Comment by ghost** [comment]
A node never knows the full path, just the next hop along it.

**Reply by faragher** [reply]
Is the tag then the number of hops? I can see the hops to the destination on the requesting system, and it would make sense to store the data to filter multiple announces and save only the shortest one.

**Reply by ThomasFreedman** [reply]
> A node never knows the full path, just the next hop along it.

Correct, according to the docs. Now that I got my code working, the path I provide is simply the app_name. It is the same value provided to the register_request_handler and then passed to the response_generator callback function when a request arrives.

I believe the path value you provide to the request handler will be the app_name + dot delimited aspects. Since I'm not using any aspects, it's just the app_name.

What this boils down to is app_name is the primary message filter, and the aspects are an optional filter within the app. 

When I started playing with the announce example it suggested to remove the aspects to see what happens. It will see every announce from all applications on RNS within "reach" !!!

As I referred back to that to write this, I see where Mark alse says:

> Destinations can also announce their existence, which will let the network 
> know they are reachable **_and automatically create paths to them..._** 

I failed to remember that when I posted [my question](https://github.com/markqvist/Reticulum/discussions/311) about paths.

**Comment by markqvist** [comment, marked as answer]
I'm reading all of your good questions here, and I am getting the sense that you've gotten some of the basics completely wrong, in regards to the functions of announces and such ;) I'll try and see if I can clear things up a bit, and in turn, questions like these also help a bunch in gauging how to improve the documentation to make the underlying concepts of Reticulum more clear, so thanks for that.

> I have successfully setup a link in response to my app's announcement. Lets call the app the "server". A link is created by a responder to the announcement.

This is *probably* not what you want to do. Imagine you have a thousand clients running on the network. Your "server" destination sends an announce. Now, all one thousand clients try to establish a link to your server at the same time. Is that what you want to do?

There *might* be situations where you want to try establishing a link in response to receiving an announce, but almost always, logic as that makes no sense. There should be something else that triggers the need to establish a link to a destination.

> Q1: Is the path request code really necessary to setup a link to the source of an announcement, or will a path to the announcer always exist?

A path to a destination *may or may not* exist at any given point. You will never know for sure without testing it (network topology could have changed, destinations dropped in or out of the network, etc). You can test this using the `Transport.has_path()` API call. If no path exists, you can request one. If one already exists, there's generally no need to try requesting one.

> Q2: What path value do I provide on the server to create a request callback handler?

Whatever you want! That is up to you :)

You're (understandably) confusing paths in *the network* with the `path` argument in the `Request` API. They're completely different things. Think of the `Request` path as the path component of an URL. It is there to help you create many different `Request` handlers, separated in a logical way.

You can register many different request handlers, each on a different path. This allows logical separation of different functions, requests and calls.

It has nothing to do with paths in the network ;)

> I also noticed the link_closed handler isn't being called on the server, even when I exit the app. I do see it on the responder, but it's not immediate, and is probably due to a timeout occuring.

This is probably because you are just killing the app, and not exiting it gracefully (with a proper `exit()` call), which would in turn run the internal exit handlers of Reticulum, and nicely tear down links and such, signalling to the other end that the link terminated.

> I'm only using one aspect element (app_name)

Don't. This is generally a bad idea. It's not that you can't do it, if you absolutely must, but it's probably a good idea to create some sort of logical separation in your application from the start.
